tests/MongoDB.Driver.Tests/AuthenticationTests.cs

```diff
@@ -294,27 +297,70 @@ public void Authentication_succeeds_when_user_has_multiple_credentials_and_mecha
             var settings = client.Settings.Clone();
             settings.Credential = MongoCredential.FromComponents(
                 mechanism: "SCRAM-SHA-256", source: null, username: uniqueUnicodeUserName, password: unicodePassword);
-            
+
             AssertAuthenticationSucceeds(settings);
         }
 #endif
-        
-        private void AssertAuthenticationSucceeds(MongoClientSettings settings)
+
+        [SkippableTheory]
+        [ParameterAttributeData]
+        public void Authentication_succeeds_with_MONGODB_X509_mechanism(
+            [Values(false, true)] bool async)
         {
-             new MongoClient(settings).ListDatabaseNames().ToEnumerable().ToList();
+            RequireEnvironment.Check().EnvironmentVariable("MONGO_X509_CLIENT_CERTIFICATE_PATH", isDefined: true);
+            RequireEnvironment.Check().EnvironmentVariable("MONGO_X509_CLIENT_CERTIFICATE_PASSWORD", isDefined: true);
+            RequireServer.Check().Tls(required: true);
+
+            var pathToClientCertificate = Environment.GetEnvironmentVariable("MONGO_X509_CLIENT_CERTIFICATE_PATH");
+            var password = Environment.GetEnvironmentVariable("MONGO_X509_CLIENT_CERTIFICATE_PASSWORD");
+            var clientCertificate = new X509Certificate2(pathToClientCertificate, password);
+
+            var userName = GetRfc2253FormattedUsernameFromX509ClientCertificate(clientCertificate);
+            var dropUserResults =
+                DropDatabaseUser(DriverTestConfiguration.Client, userName, databases: new[] { "$external" });
+            var createUserResults = CreateX509DatabaseUser(DriverTestConfiguration.Client, userName);
+
+            var settings = DriverTestConfiguration.GetClientSettings().Clone();
+            var serverVersion = CoreTestConfiguration.ServerVersion;
+            var serverExtractsUsernameFromCertificate =
+                serverVersion >= Feature.ServerExtractsUsernameFromX509Certificate.FirstSupportedVersion;
+            settings.Credential = MongoCredential.CreateMongoX509Credential(
+                username: serverExtractsUsernameFromCertificate ? null : userName);
+            settings.SslSettings = settings.SslSettings.Clone();
+            settings.SslSettings.ClientCertificates = new[] { clientCertificate };
+
+            AssertAuthenticationSucceeds(settings, async);
+        }
+
+        private void AssertAuthenticationSucceeds(MongoClientSettings settings, bool async = false)
```

Comment:
There is no point in having an async parameter here. Authentication is not affected by whether the method we call here to smoke test is async or not.

Reply:
I would refactor this method to:

    private void AssertAuthenticationSucceeds(MongoClientSettings settings)
    {
        /* If we don't use a DisposableClient, the second run of AuthenticationSucceedsWithMongoDB_X509_mechanism
         * will fail because the backing Cluster's connections will be associated with a dropped user */
        using (var client = DriverTestConfiguration.CreateDisposableclient(settings))
        {
            _ = client.ListDatabaseNames().ToList(); // will throw an exception if authentication did not succeed
        }
    }

Reply:
OK... this is very confusing, because it is not at all clear why choosing the sync vs async version of `ListDatabases` has anything to do with authentication.

Reply:
Thanks for the comment. I had not picked up on the subtle fact that a sync vs async operation indirectly triggered a sync vs async authentication.

I was assuming the connection was already authenticated and sitting in the connection pool.

Which of course could still be true... how do you know that this sync/async operation will result in a NEW connection being required (and therefore being authenticated).

Reply:
If this method needs to be called with BOTH async=false and async=true to test both authentication paths then why is the async parameter optional?

Shouldn't all callers of this method have to test both async and sync paths?

Reply:
> Which of course could still be true... how do you know that this sync/async operation will result in a NEW connection being required (and therefore being authenticated).

When I tested with breakpoints, I was able to verify that the Authenticate methods were being hit right after calling the ListDatabases command. My understanding is that the only entry point into the `Authenticate` methods is via `AuthenticateHelper.Authenticate`, which is called by `ConnectionInitializer.InitializeConnection`

Reply:
Additionally, combined with the fact that we're now using `DisposableMongoClient` and because for every scram test we're creating a new unique user (via guids) while for the X509 test we drop the user and recreate a new one, I believe it's highly unlikely that a test would use an existing connection. When this happened in earlier in this effort for the X509 test (when we were NOT using a DisposableMongoClient), the test failed on the second run because the test would attempt to use a Cluster for which the connections were no longer valid because the user associated with those connections had been dropped).